File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/ArmorTypeEnum.java
package ai.ragu.charactermanager.enumeration;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

import java.util.HashMap;
import java.util.Map;

public enum ArmorTypeEnum {
    LEATHER("leather"),
    COMPOSITE("composite"),
    CHAIN("chain"),
    PLATE("plate");

    private final String value;
    private final static Map<String, ArmorTypeEnum> CONSTANTS = new HashMap<>();

    static {
        for (ArmorTypeEnum c: values()) {
            CONSTANTS.put(c.value, c);
        }
    }

    ArmorTypeEnum(String value) {
        this.value = value;
    }

    @Override
    public String toString() {
        return this.value;
    }

    @JsonValue
    public String value() {
        return this.value;
    }

    @JsonCreator
    public static ArmorTypeEnum fromValue(String value) {
        ArmorTypeEnum constant = CONSTANTS.get(value);
        if (constant == null) {
            throw new IllegalArgumentException(value);
        } else {
            return constant;
        }
    }
}

File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/ArmorTypeEnum.kt
package ai.ragu.charactermanager.enumeration

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue
import java.util.*

enum class ArmorTypeEnum(private val value: String) {
    LEATHER("leather"), COMPOSITE("composite"), CHAIN("chain"), PLATE("plate");

    companion object {
        private val CONSTANTS: MutableMap<String, ArmorTypeEnum> = HashMap()
        @JsonCreator
        fun fromValue(value: String?): ArmorTypeEnum {
            val constant = CONSTANTS[value]
            return constant ?: throw IllegalArgumentException(value)
        }

        init {
            for (c in values()) {
                CONSTANTS[ai.ragu.charactermanager.enumeration.c.value] = ai.ragu.charactermanager.enumeration.c
            }
        }
    }

    override fun toString(): String {
        return value
    }

    @JsonValue
    fun value(): String {
        return value
    }

}